File: src/main/java/com/ptaf/ui/action_performer/AssertionPerformer.java
package com.ptaf.ui.action_performer;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.assertions.LocatorAssertions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.regex.Pattern;

import static com.microsoft.playwright.assertions.PlaywrightAssertions.assertThat;

public class AssertionPerformer {

    private static final Logger logger = LoggerFactory.getLogger(AssertionPerformer.class);

    public void performAssertion(Page page, String action, Locator targetLocator, String value) {
        try {
            switch (action.toLowerCase()) {
                // --- Text-related ---
                case "contain":
                    assertThat(targetLocator).containsText(value);
                    break;
                case "notcontain":
                    assertThat(targetLocator).not().containsText(value);
                    break;
                case "hastext":
                    assertThat(targetLocator).hasText(value);
                    break;
                case "hastextexactly":
                    assertThat(targetLocator).hasText(value, new LocatorAssertions.HasTextOptions().setUseInnerText(true));
                    break;
                case "nothastext":
                    assertThat(targetLocator).not().hasText(value);
                    break;

                // --- Visibility ---
                case "isvisible":
                    assertThat(targetLocator).isVisible();
                    break;
                case "notvisible":
                    assertThat(targetLocator).not().isVisible();
                    break;
                case "ishidden":
                    assertThat(targetLocator).isHidden();
                    break;
                case "isattached":
                    assertThat(targetLocator).isAttached();
                    break;
                case "detached":
                    assertThat(targetLocator).not().isAttached();
                    break;

                // --- State ---
                case "enabled":
                    assertThat(targetLocator).isEnabled();
                    break;
                case "disabled":
                    assertThat(targetLocator).isDisabled();
                    break;
                case "checked":
                    assertThat(targetLocator).isChecked();
                    break;
                case "notchecked":
                    assertThat(targetLocator).not().isChecked();
                    break;
                case "focused":
                    assertThat(targetLocator).isFocused();
                    break;
                case "notfocused":
                    assertThat(targetLocator).not().isFocused();
                    break;

                // --- Attributes ---
                case "hasattribute":
                    String[] attrParts = value.split("=", 2);
                    assertThat(targetLocator).hasAttribute(attrParts[0], attrParts[1]);
                    break;
                case "nothasattribute":
                    String[] notAttrParts = value.split("=", 2);
                    assertThat(targetLocator).not().hasAttribute(notAttrParts[0], notAttrParts[1]);
                    break;
                case "hasclass":
                    assertThat(targetLocator).hasClass(value);
                    break;
                case "nothasclass":
                    assertThat(targetLocator).not().hasClass(value);
                    break;
                case "hasvalue":
                    assertThat(targetLocator).hasValue(value);
                    break;

                // --- Page-level ---
                case "hastitle":
                    assertThat(page).hasTitle(value);
                    break;
                case "hasurl":
                    assertThat(page).hasURL(value);
                    break;

                // --- Collections ---
                case "hascount":
                    assertThat(targetLocator).hasCount(Integer.parseInt(value));
                    break;

                // --- Regex & Pattern Matching (Optional)
                case "matchregex":
                    assertThat(targetLocator).hasText(Pattern.compile(value));
                    break;
                case "textstartswith":
                    assertThat(targetLocator).hasText(Pattern.compile("^" + Pattern.quote(value)));
                    break;
                case "textendswith":
                    assertThat(targetLocator).hasText(Pattern.compile(Pattern.quote(value) + "$"));
                    break;

                default:
                    throw new IllegalArgumentException("Unknown Assertion: " + action);
            }
        } catch (Exception e) {
            logger.error("Error while performing assertion: {} for Target Locator {}", action, targetLocator, e);
            throw new AssertionError("Assertion failed for action: " + action + ", Locator: " + targetLocator + ", Reason: " + e.getMessage(), e);
        }
    }
}

File: ptaf/ui/assertion_performer.py
"""Playwright locator assertion dispatch (ported from AssertionPerformer.java)."""

from __future__ import annotations

import logging
import re

from playwright.sync_api import Locator, Page, expect

logger = logging.getLogger(__name__)


class AssertionPerformer:
    """Perform Playwright expect() assertions on locators and pages."""

    def perform_assertion(
        self,
        page: Page,
        action: str,
        target_locator: Locator,
        value: str | None,
    ) -> None:
        try:
            action_key = action.lower()

            match action_key:
                case "contain":
                    expect(target_locator).to_contain_text(value or "")
                case "notcontain":
                    expect(target_locator).not_to_contain_text(value or "")
                case "hastext":
                    expect(target_locator).to_have_text(value or "")
                case "hastextexactly":
                    expect(target_locator).to_have_text(
                        value or "", use_inner_text=True
                    )
                case "nothastext":
                    expect(target_locator).not_to_have_text(value or "")
                case "isvisible":
                    expect(target_locator).to_be_visible()
                case "notvisible":
                    expect(target_locator).not_to_be_visible()
                case "ishidden":
                    expect(target_locator).to_be_hidden()
                case "isattached":
                    expect(target_locator).to_be_attached()
                case "detached":
                    expect(target_locator).not_to_be_attached()
                case "enabled":
                    expect(target_locator).to_be_enabled()
                case "disabled":
                    expect(target_locator).to_be_disabled()
                case "checked":
                    expect(target_locator).to_be_checked()
                case "notchecked":
                    expect(target_locator).not_to_be_checked()
                case "focused":
                    expect(target_locator).to_be_focused()
                case "notfocused":
                    expect(target_locator).not_to_be_focused()
                case "hasattribute":
                    attr_parts = (value or "").split("=", 1)
                    expect(target_locator).to_have_attribute(
                        attr_parts[0], attr_parts[1]
                    )
                case "nothasattribute":
                    not_attr_parts = (value or "").split("=", 1)
                    expect(target_locator).not_to_have_attribute(
                        not_attr_parts[0], not_attr_parts[1]
                    )
                case "hasclass":
                    expect(target_locator).to_have_class(value or "")
                case "nothasclass":
                    expect(target_locator).not_to_have_class(value or "")
                case "hasvalue":
                    expect(target_locator).to_have_value(value or "")
                case "hastitle":
                    expect(page).to_have_title(value or "")
                case "hasurl":
                    expect(page).to_have_url(value or "")
                case "hascount":
                    expect(target_locator).to_have_count(int(value or "0"))
                case "matchregex":
                    expect(target_locator).to_have_text(re.compile(value or ""))
                case "textstartswith":
                    expect(target_locator).to_have_text(
                        re.compile("^" + re.escape(value or ""))
                    )
                case "textendswith":
                    expect(target_locator).to_have_text(
                        re.compile(re.escape(value or "") + "$")
                    )
                case _:
                    raise ValueError(f"Unknown Assertion: {action}")

        except Exception as exc:
            logger.error(
                "Error while performing assertion: %s for Target Locator %s",
                action,
                target_locator,
                exc_info=True,
            )
            raise AssertionError(
                f"Assertion failed for action: {action}, Locator: {target_locator}, "
                f"Reason: {exc}"
            ) from exc
